drivers/sftp/onitu_sftp/sftp.py

```diff
@@ -0,0 +1,325 @@
+    dirs = finalPath.split(u"/")
+
+    currentDir = u""
+    doStat = True
+    try:
+        while dirs:
+            currentDir += dirs.pop(0) + u"/"
+            if doStat:
+                try:
+                    sftp.stat(currentDir)
+                # The current dir doesn't exists, so we create it
+                except IOError as ioe:
+                    if ioe.errno == 2:
+                        sftp.mkdir(currentDir)
+                        # If a dir was not found, it means all the next
+                        # won't exist either. So stop bothering to stat
+                        doStat = False
+                    else:
+                        raise
+            else:
+                sftp.mkdir(currentDir)
+    except IOError as ioe:
+        raise ServiceError(u"Error while creating subdirs {}: {}"
+                           .format(finalPath, ioe))
+
+
+def update_metadata(metadata, size, mtime, update_file=False):
+    metadata.size = size
+    metadata.extra['mtime'] = mtime
+    metadata.write()
+    if update_file:
+        plug.update_file(metadata)
+
+
+@plug.handler()
+def get_file(metadata):
+    plug.logger.debug(u"Getting remote file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        f = sftp.open(metadata.path, 'r')
+        # "If reading the entire file, pre-fetching can dramatically improve
+        # the download speed by avoiding roundtrip latency."
+        # http://docs.paramiko.org/en/latest/api/sftp.html
+        # #paramiko.sftp_file.SFTPFile.prefetch
+        f.prefetch()
+        data = f.read()
+        return data
+    except IOError as e:
+        raise ServiceError("Error getting file '{}': {}"
+                           .format(metadata.path, e))
+    finally:
+        sftp.close()
+
+
+@plug.handler()
+def start_upload(metadata):
+    plug.logger.debug(u"Starting upload of {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        create_file_subdirs(sftp, metadata.path)
+        plug.logger.debug(u"Adding {} to ignored files".format(metadata.path))
+        events_to_ignore.add(metadata.path)
+    except ServiceError:
+        raise
+    finally:
+        sftp.close()
+
+
+@plug.handler()
+def upload_file(metadata, content):
+    plug.logger.debug(u"Uploading file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        f = sftp.open(metadata.path, 'w+')
+        f.write(content)
+        f.close()
+    except (IOError, OSError) as e:
+        raise ServiceError("Error writing file '{}': {}"
+                           .format(metadata.path, e))
+    finally:
+        sftp.close()
+
+
+@plug.handler()
+def end_upload(metadata):
+    plug.logger.debug(u"Ending upload of file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
+    try:
+        stat_res = sftp.stat(metadata.path)
+        update_metadata(metadata, stat_res.st_size, stat_res.st_mtime,
```

Comment:
You should *not* update the size at the end of an upload, this is already known and could confuse other services if the one returned by SFTP is wrong.